### PR TITLE
Update didyoumean for Komada 0.20.x

### PR DESCRIPTION
**Pull request by CyberiumShadow**

### CODE REVIEW

**Comment by CyberiumShadow** [comment]
This should fix #80

File: monitors/didyoumean.js [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
 exports.run = async (client, msg) => {
   if (msg.author.bot) return;
   if (client.config.selfbot && msg.author.id !== client.user.id) return;
 
-  const conf = client.configuration.get(msg.guild);
-  const prefixLength = client.funcs.parseCommand(client, msg, true);
+  const { prefixLength, command, prefix } = parseCommand(client, msg);
   if (!prefixLength) return;
-  const command = client.funcs.parseCommand(client, msg);
   if (command.length && !(client.commands.has(command) || client.aliases.has(command))) {
```
Comment (kyranet) [inline monitors/didyoumean.js]: Let's make this a negative check, so we don't have a christmas tree.

File: monitors/didyoumean.js [hunk 2]
```diff
@@ -1,13 +1,13 @@
+/* eslint-disable no-use-before-define */
+
 const levenshtein = require("fast-levenshtein");
 
 exports.run = async (client, msg) => {
   if (msg.author.bot) return;
```
Comment (kyranet) [inline monitors/didyoumean.js]: Instead of two ifs, you could merge both in one.

File: monitors/didyoumean.js [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
   if (client.config.selfbot && msg.author.id !== client.user.id) return;
 
-  const conf = client.configuration.get(msg.guild);
-  const prefixLength = client.funcs.parseCommand(client, msg, true);
+  const { prefixLength, command, prefix } = parseCommand(client, msg);
   if (!prefixLength) return;
-  const command = client.funcs.parseCommand(client, msg);
   if (command.length && !(client.commands.has(command) || client.aliases.has(command))) {
     const distances = [];
     client.commands.filter(c => c.conf.permLevel <= msg.member.permLevel).forEach((val, cmd) => distances.push({
```
Comment (kyranet) [inline monitors/didyoumean.js]: Use a for-of instead here, to iterate over commands once.

File: monitors/didyoumean.js [hunk 4]
```diff
@@ -16,8 +16,8 @@ exports.run = async (client, msg) => {
     }));
     distances.sort((a, b) => (a.score < b.score ? 1 : -1));
```
Comment (kyranet) [inline monitors/didyoumean.js]: Check if the array's length is at least one, before trying to sort it.

File: monitors/didyoumean.js [hunk 5]
```diff
@@ -16,8 +16,8 @@ exports.run = async (client, msg) => {
...
     if (distances[0] && distances[0].dist <= 1) {
-      const message = await msg.channel.send(`|\`❔\`| Did you mean \`${conf.prefix + distances[0].cmd}\`?`);
-      setTimeout(() => { if (message.deletable) message.delete(); }, 10000);
+      const message = await msg.channel.send(`|\`❔\`| Did you mean \`${prefix + distances[0].cmd}\`?`);
```
Comment (kyranet) [inline monitors/didyoumean.js]: Use `msg.send` instead.

File: monitors/didyoumean.js [hunk 6]
```diff
@@ -16,8 +16,8 @@ exports.run = async (client, msg) => {
...
+      if (message.deletable) message.delete({ timeout: 10000 });
```
Comment (kyranet) [inline monitors/didyoumean.js]: message.deletable becomes redundant there. You should await and catch that message.delete.

File: monitors/didyoumean.js [hunk 7]
```diff
@@ -16,8 +16,8 @@ exports.run = async (client, msg) => {
     }));
     distances.sort((a, b) => (a.score < b.score ? 1 : -1));
     if (distances[0] && distances[0].dist <= 1) {
```
Comment (kyranet) [inline monitors/didyoumean.js]: Make the minimum distance a constant/clientConfig variable.

**Review by kyranet — CHANGES_REQUESTED** [review]

File: monitors/didyoumean.js [hunk 8]
```diff
@@ -1,24 +1,33 @@
...
-    if (distances[0] && distances[0].dist <= 1) {
-      const message = await msg.channel.send(`|\`❔\`| Did you mean \`${conf.prefix + distances[0].cmd}\`?`);
-      setTimeout(() => { if (message.deletable) message.delete(); }, 10000);
-    }
+    });
+  }
+
+  if (distances.length === 0) return;
+  distances.sort((a, b) => (a.score < b.score ? 1 : -1));
+  if (distances[0] && distances[0].dist <= this.minDist) {
```
Comment (kyranet) [inline monitors/didyoumean.js]: We know the first object in distances exists, there's a check before.

File: monitors/didyoumean.js [hunk 9]
```diff
@@ -1,24 +1,33 @@
...
+    await msg.send(`|\`❔\`| Did you mean \`${prefix + distances[0].cmd}\`?`);
```
Comment (kyranet) [inline monitors/didyoumean.js]: return and catch it.

File: monitors/didyoumean.js [hunk 10]
```diff
@@ -1,24 +1,33 @@
...
-  const command = client.funcs.parseCommand(client, msg);
-  if (command.length && !(client.commands.has(command) || client.aliases.has(command))) {
-    const distances = [];
-    client.commands.filter(c => c.conf.permLevel <= msg.member.permLevel).forEach((val, cmd) => distances.push({
+  if (!command.length && (client.commands.has(command) || client.aliases.has(command))) return;
+
+  const distances = [];
+
+  for (const cmd in client.commands) {
+    if (!client.funcs.checkPerms(client, msg, cmd.conf.permLevel)) continue;
```
Comment (kyranet) [inline monitors/didyoumean.js]: This is okay for a selfbot, but that's almost like generating the help command in every single message that starts with the prefix, make an external function able to calculate highest permission level and try again.

**Review by kyranet — DISMISSED** [review]
More things

File: monitors/didyoumean.js [hunk 11]
```diff
@@ -18,8 +18,10 @@ exports.run = async (client, msg) => {
 
   const distances = [];
 
+  const highPerm = calcHighPerm(client, msg);
+
   for (const cmd in client.commands) {
```
Comment (UnseenFaith) [inline monitors/didyoumean.js]: remove this and use msg.usableCommands extended property. Does exactly what you want.

lets pretend i clicked request changes and not comment here.

Comment (UnseenFaith) [reply monitors/didyoumean.js]: another reason why this should be changed: 
usablecommands runs all of the inhibitors, while the previous code doesn't, so you could accidentially expose hidden commands or something similar

So if you don't want to use usableCommands, just make sure you runInhibitors before validating that the command was found

File: monitors/didyoumean.js [hunk 12]
```diff
@@ -1,24 +1,38 @@
-const levenshtein = require("fast-levenshtein");
+/* eslint-disable no-use-before-define, no-restricted-syntax, consistent-return, guard-for-in */
+
+let levenshtein;
+
+exports.minDist = null;
+
+exports.init = (client) => {
+  levenshtein = require("fast-levenshtein");
```
Comment (kyranet) [inline monitors/didyoumean.js]: ????? This is Komada 0.20.x, not 0.18.1, the download command supports packages.

**Review by kyranet — DISMISSED** [review]

File: monitors/didyoumean.js (hunk 12 above)
Comment (CyberiumShadow) [reply monitors/didyoumean.js]: It didn't when i was testing it initially.

Kept firing `cannot find module fast-levenshtein`

Comment (CyberiumShadow) [reply monitors/didyoumean.js]: probs just a weird issue with my docker. anyways reverted that line to be outside init

**Review by kyranet — APPROVED** [review]
LGTM.

**Pull request merged**